cmd/edgeql-go/gen.go

```diff
@@ -0,0 +1,334 @@
+// This source file is part of the EdgeDB open source project.
+//
+// Copyright 2020-present EdgeDB Inc. and the EdgeDB authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+
+	"github.com/edgedb/edgedb-go/internal/codecs"
+	"github.com/edgedb/edgedb-go/internal/descriptor"
+	types "github.com/edgedb/edgedb-go/internal/edgedbtypes"
+)
+
+type lookupKey struct {
+	ID       types.UUID
+	Required bool
+}
+
+// Type is a go type definition
+type Type struct {
+	definition string
+	imports    []string
+}
+
+// Generator generates go code from EdgeDB types.
+type Generator struct {
+	mx             sync.RWMutex
+	typeNameLookup map[lookupKey]Type
+}
+
+func (g *Generator) getType(
+	desc descriptor.Descriptor,
+	required bool,
+) (Type, error) {
+	key := lookupKey{desc.ID, required}
+
+	g.mx.RLock()
+	typ, ok := g.typeNameLookup[key]
+	g.mx.RUnlock()
+
+	if !ok {
+		text, imports, err := g.generateType(desc, required)
+		if err != nil {
+			return Type{}, err
+		}
+
+		typ = Type{string(text), imports}
+
+		g.mx.Lock()
+		g.typeNameLookup[key] = typ
+		g.mx.Unlock()
+	}
+
+	return typ, nil
+}
+
+func (g *Generator) generateType(
+	desc descriptor.Descriptor,
+	required bool,
+) ([]byte, []string, error) {
+	var (
+		buf     bytes.Buffer
+		err     error
+		imports []string
+	)
+
+	switch desc.Type {
+	case descriptor.Set, descriptor.Array:
+		imports, err = g.generateSlice(&buf, desc)
+	case descriptor.Object, descriptor.NamedTuple:
+		imports, err = g.generateObject(&buf, desc, required)
+	case descriptor.Tuple:
+		imports, err = g.generateTuple(&buf, desc, required)
+	case descriptor.BaseScalar, descriptor.Scalar, descriptor.Enum:
```

Comment:
`Scalar` and `Enum` have non-built-in type descriptor IDs, `generateBaseScalar()` will generate a broken Go file currently.

Reply:
fixed